speechbrain/processing/features.py

```diff
@@ -1316,11 +1286,92 @@ def _load(self, path, end_of_epoch=False):
+    """Welford's algorithm for running mean, variance (from Wikipedia)
+
+    WARNING: Must be called in sync across processes.
+
+    Arguments
+    ---------
+    x : torch.Tensor
+        The new values to add to the running stats.
+    mask : torch.Tensor
+        A mask indicating padding positions to be ignored.
+    dim : tuple or int
+        The dimension or dimensions to reduce (e.g. 1 for length).
+    run_count : float or torch.Tensor
+        The running number of samples seen so far.
+    run_mean : float or torch.Tensor
+        The running mean of samples seen so far.
+    run_std : float or torch.Tensor, optional
+        The running standard deviations from the mean. If `None`, the
+        variance is not computed and this returns 1.0 for the `run_std`.
+
+    Returns
+    -------
+    run_count : torch.Tensor
+        Updated count all samples so far.
+    run_mean : torch.Tensor
+        Updated running mean of all samples so far.
+    run_std : torch.Tensor (if passed)
+        Updated running standard deviations of all samples so far.
+    """
+    delta = x - run_mean
+    n = ddp_all_reduce(mask.sum(dim), ReduceOp.SUM)
+    delta_sum = ddp_all_reduce((delta * mask).sum(dim), ReduceOp.SUM)
+    run_mean += delta_sum / (run_count + n)
+
+    # Only compute variance if needed (for efficiency)
+    if run_std is not None:
+        var_sum = (delta * (x - run_mean) * mask).sum(dim)
+        new_var = ddp_all_reduce(var_sum, ReduceOp.SUM) / (n - 1)
+        run_var = (run_std.square() * run_count + new_var * n) / (run_count + n)
+        run_std = run_var.sqrt()
+    else:
+        run_std = torch.ones_like(run_mean)
+
+    return run_count + n.sum(), run_mean, run_std
+
+
+def get_mask(x, lengths=None, length_dim=1):
+    """Create a mask from relative lengths along a given dimension.
+
+    Arguments
+    ---------
+    x : torch.Tensor
+        The input tensor demonstrating the size of the target mask.
+    lengths : torch.Tensor, optional
+        The relative lengths of an input batch of utterances.
+    length_dim : int, default 1
+        The dimension for which the lengths indicate padded positions.
+
+    Returns
+    -------
+    mask : torch.Tensor
+        A boolean tensor with `True` for valid positions and `False`
+        for padding positions.
+    """
+    if lengths is None:
+        return torch.ones_like(x)
+
+    # Convert relative lengths to absolute lengths, then compute boolean mask
+    max_len = x.size(length_dim)
+    abs_lengths = lengths.unsqueeze(1) * max_len
+    mask = torch.arange(max_len, device=x.device).unsqueeze(0) < abs_lengths
+
+    # Add dimensions other than (batch, length) back into the mask
+    for dim in range(1, x.ndim):
+        if dim != length_dim:
+            mask = mask.unsqueeze(dim)
+
+    # Repeat values (without copying) into singleton dimensions
+    return mask.expand_as(x)
+
+
 class GlobalNorm(torch.nn.Module):
```

Comment:
Was this function wrong as well? I see it being used in AudioMnist.

Reply:
Yes, it used the same wrong std update algorithm.